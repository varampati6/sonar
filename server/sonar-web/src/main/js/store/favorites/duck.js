/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import { uniq, without } from 'lodash';

type Favorite = { key: string };

type ReceiveFavoritesAction = {
  type: 'RECEIVE_FAVORITES',
  favorites: Array<Favorite>,
  notFavorites: Array<Favorite>
};

type AddFavoriteAction = {
  type: 'ADD_FAVORITE',
  componentKey: string
};

type RemoveFavoriteAction = {
  type: 'REMOVE_FAVORITE',
  componentKey: string
};

type Action = ReceiveFavoritesAction | AddFavoriteAction | RemoveFavoriteAction;

type State = Array<string>;

export const actions = {
  RECEIVE_FAVORITES: 'RECEIVE_FAVORITES',
  ADD_FAVORITE: 'ADD_FAVORITE',
  REMOVE_FAVORITE: 'REMOVE_FAVORITE'
};

export const receiveFavorites = (
  favorites: Array<Favorite>,
  notFavorites: Array<Favorite> = []
): ReceiveFavoritesAction => ({
  type: actions.RECEIVE_FAVORITES,
  favorites,
  notFavorites
});

export const addFavorite = (componentKey: string): AddFavoriteAction => ({
  type: actions.ADD_FAVORITE,
  componentKey
});

export const removeFavorite = (componentKey: string): RemoveFavoriteAction => ({
  type: actions.REMOVE_FAVORITE,
  componentKey
});

export default (state: State = [], action: Action): State => {
  if (action.type === actions.RECEIVE_FAVORITES) {
    const toAdd = action.favorites.map(f => f.key);
    const toRemove = action.notFavorites.map(f => f.key);
    return without(uniq([...state, ...toAdd]), ...toRemove);
  }

  if (action.type === actions.ADD_FAVORITE) {
    return uniq([...state, action.componentKey]);
  }

  if (action.type === actions.REMOVE_FAVORITE) {
    return without(state, action.componentKey);
  }

  return state;
};

export const isFavorite = (state: State, componentKey: string) => state.includes(componentKey);
