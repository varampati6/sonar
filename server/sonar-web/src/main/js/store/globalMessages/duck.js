/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import { uniqueId } from 'lodash';

type Level = 'ERROR' | 'SUCCESS';

type Message = {
  id: string,
  message: string,
  level: Level
};

export type State = Array<Message>;

type Action = Object;

export const ERROR = 'ERROR';
export const SUCCESS = 'SUCCESS';

/* Actions */
const ADD_GLOBAL_MESSAGE = 'ADD_GLOBAL_MESSAGE';
const CLOSE_GLOBAL_MESSAGE = 'CLOSE_GLOBAL_MESSAGE';
const CLOSE_ALL_GLOBAL_MESSAGES = 'CLOSE_ALL_GLOBAL_MESSAGES';

const addGlobalMessageActionCreator = (id: string, message: string, level: Level) => ({
  type: ADD_GLOBAL_MESSAGE,
  message,
  level,
  id
});

export const closeGlobalMessage = (id: string) => ({
  type: CLOSE_GLOBAL_MESSAGE,
  id
});

export const closeAllGlobalMessages = (id: string) => ({
  type: CLOSE_ALL_GLOBAL_MESSAGES,
  id
});

const addGlobalMessage = (message: string, level: Level) => (dispatch: Function) => {
  const id = uniqueId('global-message-');
  dispatch(addGlobalMessageActionCreator(id, message, level));
  setTimeout(() => dispatch(closeGlobalMessage(id)), 5000);
};

export const addGlobalErrorMessage = (message: string) => addGlobalMessage(message, ERROR);

export const addGlobalSuccessMessage = (message: string) => addGlobalMessage(message, SUCCESS);

/* Reducer */
const globalMessages = (state: State = [], action: Action = {}) => {
  switch (action.type) {
    case ADD_GLOBAL_MESSAGE:
      return [
        {
          id: action.id,
          message: action.message,
          level: action.level
        }
      ];

    case 'REQUIRE_AUTHORIZATION':
      // FIXME l10n
      return [
        {
          id: uniqueId('global-message-'),
          message: 'You are not authorized to access this page. ' +
            'Please log in with more privileges and try again.',
          level: ERROR
        }
      ];

    case CLOSE_GLOBAL_MESSAGE:
      return state.filter(message => message.id !== action.id);

    case CLOSE_ALL_GLOBAL_MESSAGES:
      return [];
    default:
      return state;
  }
};

export default globalMessages;

/* Selectors */
export const getGlobalMessages = (state: State) => state;
