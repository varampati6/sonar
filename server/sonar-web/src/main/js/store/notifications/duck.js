/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import { combineReducers } from 'redux';
import { uniqBy, uniqWith } from 'lodash';

export type Notification = {
  channel: string,
  type: string,
  project?: string,
  projectName?: string,
  organization?: string
};

export type NotificationsState = Array<Notification>;
export type ChannelsState = Array<string>;
export type TypesState = Array<string>;

type AddNotificationAction = {
  type: 'ADD_NOTIFICATION',
  notification: Notification
};

type RemoveNotificationAction = {
  type: 'REMOVE_NOTIFICATION',
  notification: Notification
};

type ReceiveNotificationsAction = {
  type: 'RECEIVE_NOTIFICATIONS',
  notifications: NotificationsState,
  channels: ChannelsState,
  globalTypes: TypesState,
  perProjectTypes: TypesState
};

type Action = AddNotificationAction | RemoveNotificationAction | ReceiveNotificationsAction;

export const addNotification = (notification: Notification): AddNotificationAction => ({
  type: 'ADD_NOTIFICATION',
  notification
});

export const removeNotification = (notification: Notification): RemoveNotificationAction => ({
  type: 'REMOVE_NOTIFICATION',
  notification
});

export const receiveNotifications = (
  notifications: NotificationsState,
  channels: ChannelsState,
  globalTypes: TypesState,
  perProjectTypes: TypesState
): ReceiveNotificationsAction => ({
  type: 'RECEIVE_NOTIFICATIONS',
  notifications,
  channels,
  globalTypes,
  perProjectTypes
});

const onAddNotification = (state: NotificationsState, notification: Notification) => {
  const isNotificationsEqual = (a: Notification, b: Notification) =>
    a.channel === b.channel && a.type === b.type && a.project === b.project;
  return uniqWith([...state, notification], isNotificationsEqual);
};

const onRemoveNotification = (state: NotificationsState, notification: Notification) => {
  return state.filter(
    n =>
      n.channel !== notification.channel ||
      n.type !== notification.type ||
      n.project !== notification.project
  );
};

const onReceiveNotifications = (state: NotificationsState, notifications: NotificationsState) => {
  return [...notifications];
};

const notifications = (state: NotificationsState = [], action: Action) => {
  switch (action.type) {
    case 'ADD_NOTIFICATION':
      return onAddNotification(state, action.notification);
    case 'REMOVE_NOTIFICATION':
      return onRemoveNotification(state, action.notification);
    case 'RECEIVE_NOTIFICATIONS':
      return onReceiveNotifications(state, action.notifications);
    default:
      return state;
  }
};

const channels = (state: ChannelsState = [], action: Action) => {
  if (action.type === 'RECEIVE_NOTIFICATIONS') {
    return action.channels;
  } else {
    return state;
  }
};

const globalTypes = (state: TypesState = [], action: Action) => {
  if (action.type === 'RECEIVE_NOTIFICATIONS') {
    return action.globalTypes;
  } else {
    return state;
  }
};

const perProjectTypes = (state: TypesState = [], action: Action) => {
  if (action.type === 'RECEIVE_NOTIFICATIONS') {
    return action.perProjectTypes;
  } else {
    return state;
  }
};

type State = {
  notifications: NotificationsState,
  channels: ChannelsState,
  globalTypes: TypesState,
  perProjectTypes: TypesState
};

export default combineReducers({ notifications, channels, globalTypes, perProjectTypes });

export const getGlobal = (state: State): NotificationsState =>
  state.notifications.filter(n => !n.project);

export const getProjects = (state: State): Array<{ key: string, name: string }> => {
  // $FlowFixMe
  const allProjects = state.notifications.filter(n => n.project != null).map(n => ({
    key: n.project,
    name: n.projectName,
    organization: n.organization
  }));

  return uniqBy(allProjects, project => project.key);
};

export const getForProject = (state: State, project: string): NotificationsState =>
  state.notifications.filter(n => n.project === project);

export const getChannels = (state: State): ChannelsState => state.channels;

export const getGlobalTypes = (state: State): TypesState => state.globalTypes;

export const getPerProjectTypes = (state: State): TypesState => state.perProjectTypes;
