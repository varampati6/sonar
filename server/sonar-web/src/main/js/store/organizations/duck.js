/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import { combineReducers } from 'redux';
import { omit, uniq, without } from 'lodash';

export type Organization = {
  adminPages?: Array<{ key: string, name: string }>,
  avatar?: string,
  canAdmin?: boolean,
  canDelete?: boolean,
  canProvisionProjects?: boolean,
  canUpdateProjectsVisibilityToPrivate?: boolean,
  description?: string,
  key: string,
  name: string,
  pages?: Array<{ key: string, name: string }>,
  projectVisibility: string,
  url?: string
};

export type OrgGroup = {
  id: string,
  default: boolean,
  description: string,
  membersCount: number,
  name: string
};

type ReceiveOrganizationsAction = {
  type: 'RECEIVE_ORGANIZATIONS',
  organizations: Array<Organization>
};

type ReceiveMyOrganizationsAction = {
  type: 'RECEIVE_MY_ORGANIZATIONS',
  organizations: Array<Organization>
};

type ReceiveOrganizationGroups = {
  type: 'RECEIVE_ORGANIZATION_GROUPS',
  key: string,
  groups: Array<OrgGroup>
};

type CreateOrganizationAction = {
  type: 'CREATE_ORGANIZATION',
  organization: Organization
};

type UpdateOrganizationAction = {
  type: 'UPDATE_ORGANIZATION',
  key: string,
  changes: {}
};

type DeleteOrganizationAction = {
  type: 'DELETE_ORGANIZATION',
  key: string
};

type Action =
  | ReceiveOrganizationsAction
  | ReceiveMyOrganizationsAction
  | ReceiveOrganizationGroups
  | CreateOrganizationAction
  | UpdateOrganizationAction
  | DeleteOrganizationAction;

type ByKeyState = {
  [key: string]: Organization
};

type GroupsState = {
  [key: string]: Array<OrgGroup>
};

type MyState = Array<string>;

type State = {
  byKey: ByKeyState,
  my: MyState,
  groups: GroupsState
};

export const receiveOrganizations = (
  organizations: Array<Organization>
): ReceiveOrganizationsAction => ({
  type: 'RECEIVE_ORGANIZATIONS',
  organizations
});

export const receiveMyOrganizations = (
  organizations: Array<Organization>
): ReceiveMyOrganizationsAction => ({
  type: 'RECEIVE_MY_ORGANIZATIONS',
  organizations
});

export const receiveOrganizationGroups = (
  key: string,
  groups: Array<OrgGroup>
): receiveOrganizationGroups => ({
  type: 'RECEIVE_ORGANIZATION_GROUPS',
  key,
  groups
});

export const createOrganization = (organization: Organization): CreateOrganizationAction => ({
  type: 'CREATE_ORGANIZATION',
  organization
});

export const updateOrganization = (key: string, changes: {}): UpdateOrganizationAction => ({
  type: 'UPDATE_ORGANIZATION',
  key,
  changes
});

export const deleteOrganization = (key: string): DeleteOrganizationAction => ({
  type: 'DELETE_ORGANIZATION',
  key
});

const onReceiveOrganizations = (
  state: ByKeyState,
  action: ReceiveOrganizationsAction | ReceiveMyOrganizationsAction
): ByKeyState => {
  const nextState = { ...state };
  action.organizations.forEach(organization => {
    nextState[organization.key] = { ...state[organization.key], ...organization };
  });
  return nextState;
};

const byKey = (state: ByKeyState = {}, action: Action) => {
  switch (action.type) {
    case 'RECEIVE_ORGANIZATIONS':
    case 'RECEIVE_MY_ORGANIZATIONS':
      return onReceiveOrganizations(state, action);
    case 'CREATE_ORGANIZATION':
      return { ...state, [action.organization.key]: action.organization };
    case 'UPDATE_ORGANIZATION':
      return {
        ...state,
        [action.key]: {
          ...state[action.key],
          key: action.key,
          ...action.changes
        }
      };
    case 'DELETE_ORGANIZATION':
      return omit(state, action.key);
    default:
      return state;
  }
};

const my = (state: MyState = [], action: Action) => {
  switch (action.type) {
    case 'RECEIVE_MY_ORGANIZATIONS':
      return uniq([...state, ...action.organizations.map(o => o.key)]);
    case 'CREATE_ORGANIZATION':
      return uniq([...state, action.organization.key]);
    case 'DELETE_ORGANIZATION':
      return without(state, action.key);
    default:
      return state;
  }
};

const groups = (state: GroupsState = {}, action: Action) => {
  switch (action.type) {
    case 'RECEIVE_ORGANIZATION_GROUPS':
      return { ...state, [action.key]: action.groups };
    default:
      return state;
  }
};

export default combineReducers({ byKey, my, groups });

export const getOrganizationByKey = (state: State, key: string): Organization => state.byKey[key];

export const getOrganizationGroupsByKey = (state: State, key: string): Array<OrgGroup> =>
  state.groups[key] || [];

export const getMyOrganizations = (state: State): Array<Organization> =>
  state.my.map(key => getOrganizationByKey(state, key));

export const areThereCustomOrganizations = (state: State): boolean =>
  Object.keys(state.byKey).length > 1;
