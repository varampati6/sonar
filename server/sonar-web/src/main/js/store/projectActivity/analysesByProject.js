/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import type { Action, ReceiveProjectActivityAction, DeleteAnalysisAction } from './duck';

export type State = {
  [key: string]: Array<string>
};

const receiveProjectActivity = (state: State, action: ReceiveProjectActivityAction): State => {
  const analyses = state[action.project] || [];
  const newAnalyses = action.analyses.map(analysis => analysis.key);
  return {
    ...state,
    [action.project]: action.paging.pageIndex === 1 ? newAnalyses : [...analyses, ...newAnalyses]
  };
};

const deleteAnalysis = (state: State, action: DeleteAnalysisAction): State => {
  const analyses = state[action.project];
  return {
    ...state,
    [action.project]: analyses.filter(key => key !== action.analysis)
  };
};

export default (state: State = {}, action: Action): State => {
  switch (action.type) {
    case 'RECEIVE_PROJECT_ACTIVITY':
      return receiveProjectActivity(state, action);
    case 'DELETE_PROJECT_ACTIVITY_ANALYSIS':
      return deleteAnalysis(state, action);
    default:
      return state;
  }
};
