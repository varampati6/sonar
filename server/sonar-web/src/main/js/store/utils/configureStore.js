/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';

const middlewares = [thunk];
const composed = [];

if (process.env.NODE_ENV !== 'production') {
  const createLogger = require('redux-logger');
  middlewares.push(createLogger());

  composed.push(window.devToolsExtension ? window.devToolsExtension() : f => f);
}

const finalCreateStore = compose(applyMiddleware(...middlewares), ...composed)(createStore);

export default function configureStore(rootReducer, initialState) {
  return finalCreateStore(rootReducer, initialState);
}

export const configureTestStore = (rootReducer, initialState) =>
  createStore(rootReducer, initialState);
