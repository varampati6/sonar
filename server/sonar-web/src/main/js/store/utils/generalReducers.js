/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// Author: Christoffer Niska <christofferniska@gmail.com>
// https://gist.github.com/crisu83/42ecffccad9d04c74605fbc75c9dc9d1
import { uniq } from 'lodash';

/**
 * Creates a reducer that manages a single value.
 *
 * @param {function(state, action)} shouldUpdate
 * @param {function(state, action)} shouldReset
 * @param {function(state, action)} getValue
 * @param {*} defaultValue
 * @returns {function(state, action)}
 */
export const createValue = (
  shouldUpdate = () => true,
  shouldReset = () => false,
  getValue = (state, action) => action.payload,
  defaultValue = null
) => (state = defaultValue, action = {}) => {
  if (shouldReset(state, action)) {
    return defaultValue;
  }
  if (shouldUpdate(state, action)) {
    return getValue(state, action);
  }
  return state;
};

/**
 * Creates a reducer that manages a map.
 *
 * @param {function(state, action)} shouldUpdate
 * @param {function(state, action)} shouldReset
 * @param {function(state, action)} getValues
 * @returns {function(state, action)}
 */
export const createMap = (
  shouldUpdate = () => true,
  shouldReset = () => false,
  getValues = (state, action) => action.payload
) =>
  createValue(
    shouldUpdate,
    shouldReset,
    (state, action) => ({ ...state, ...getValues(state, action) }),
    {}
  );

/**
 * Creates a reducer that manages a set.
 *
 * @param {function(state, action)} shouldUpdate
 * @param {function(state, action)} shouldReset
 * @param {function(state, action)} getValues
 * @returns {function(state, action)}
 */
export const createSet = (
  shouldUpdate = () => true,
  shouldReset = () => false,
  getValues = (state, action) => action.payload
) =>
  createValue(
    shouldUpdate,
    shouldReset,
    (state, action) => uniq([...state, ...getValues(state, action)]),
    []
  );

/**
 * Creates a reducer that manages a flag.
 *
 * @param {function(state, action)} shouldTurnOn
 * @param {function(state, action)} shouldTurnOff
 * @param {bool} defaultValue
 * @returns {function(state, action)}
 */
export const createFlag = (shouldTurnOn, shouldTurnOff, defaultValue = false) => (
  state = defaultValue,
  action = {}
) => {
  if (shouldTurnOn(state, action)) {
    return true;
  }
  if (shouldTurnOff(state, action)) {
    return false;
  }
  return state;
};
