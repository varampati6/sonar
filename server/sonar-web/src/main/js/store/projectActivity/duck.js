/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import { combineReducers } from 'redux';
import analyses, * as fromAnalyses from './analyses';
import type { State as AnalysesState } from './analyses';
import analysesByProject from './analysesByProject';
import type { State as AnalysesByProjectState } from './analysesByProject';
import events, * as fromEvents from './events';
import type { State as EventsState } from './events';
import paging from './paging';
import type { State as PagingState } from './paging';

export type Event = {
  key: string,
  name: string,
  category: string,
  description?: string
};

export type Analysis = {
  key: string,
  date: string,
  events: Array<Event>
};

export type Paging = {
  total: number,
  pageIndex: number,
  pageSize: number
};

export type ReceiveProjectActivityAction = {
  type: 'RECEIVE_PROJECT_ACTIVITY',
  project: string,
  analyses: Array<Analysis>,
  paging: Paging
};

export type AddEventAction = {
  type: 'ADD_PROJECT_ACTIVITY_EVENT',
  analysis: string,
  event: Event
};

export type DeleteEventAction = {
  type: 'DELETE_PROJECT_ACTIVITY_EVENT',
  analysis: string,
  event: string
};

export type ChangeEventAction = {
  type: 'CHANGE_PROJECT_ACTIVITY_EVENT',
  event: string,
  changes: Object
};

export type DeleteAnalysisAction = {
  type: 'DELETE_PROJECT_ACTIVITY_ANALYSIS',
  project: string,
  analysis: string
};

export type Action =
  | ReceiveProjectActivityAction
  | AddEventAction
  | DeleteEventAction
  | ChangeEventAction
  | DeleteAnalysisAction;

export const receiveProjectActivity = (
  project: string,
  analyses: Array<Analysis>,
  paging: Paging
): ReceiveProjectActivityAction => ({
  type: 'RECEIVE_PROJECT_ACTIVITY',
  project,
  analyses,
  paging
});

export const addEvent = (analysis: string, event: Event): AddEventAction => ({
  type: 'ADD_PROJECT_ACTIVITY_EVENT',
  analysis,
  event
});

export const deleteEvent = (analysis: string, event: string): DeleteEventAction => ({
  type: 'DELETE_PROJECT_ACTIVITY_EVENT',
  analysis,
  event
});

export const changeEvent = (event: string, changes: Object): ChangeEventAction => ({
  type: 'CHANGE_PROJECT_ACTIVITY_EVENT',
  event,
  changes
});

export const deleteAnalysis = (project: string, analysis: string): DeleteAnalysisAction => ({
  type: 'DELETE_PROJECT_ACTIVITY_ANALYSIS',
  project,
  analysis
});

type State = {
  analyses: AnalysesState,
  analysesByProject: AnalysesByProjectState,
  events: EventsState,
  filter: string,
  paging: PagingState
};

export default combineReducers({ analyses, analysesByProject, events, paging });

const getEvent = (state: State, key: string): Event => fromEvents.getEvent(state.events, key);

const getAnalysis = (state: State, key: string) => {
  const analysis = fromAnalyses.getAnalysis(state.analyses, key);
  const events: Array<Event> = analysis.events.map(key => getEvent(state, key));
  return { ...analysis, events };
};

export const getAnalyses = (state: State, project: string) =>
  state.analysesByProject[project] &&
  state.analysesByProject[project].map(key => getAnalysis(state, key));
export const getPaging = (state: State, project: string) => state.paging[project];
