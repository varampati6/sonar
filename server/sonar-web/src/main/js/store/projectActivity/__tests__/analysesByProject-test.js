/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { configureTestStore } from '../../utils/configureStore';
import analysesByProject from '../analysesByProject';
import { receiveProjectActivity, deleteAnalysis } from '../duck';

const PROJECT_FOO = 'project-foo';
const PROJECT_BAR = 'project-bar';

const ANALYSES_FOO = [
  {
    key: 'AVgFqeOSKpGuA48ADATE',
    date: '2016-10-27T12:21:15+0200',
    events: []
  },
  {
    key: 'AVgAgC1Vdo07z3PUnnkt',
    date: '2016-10-26T12:17:29+0200',
    events: []
  }
];

const ANALYSES_FOO_2 = [
  {
    key: 'AVgFqeOSKpGuA48ADATX',
    date: '2016-10-27T12:21:15+0200',
    events: []
  }
];

const ANALYSES_BAR = [
  {
    key: 'AVgGkRvCrrTJiPpCD-rG',
    date: '2016-10-27T16:33:50+0200',
    events: []
  }
];

const PAGING = {
  total: 3,
  pageIndex: 1,
  pageSize: 100
};

it('reducer', () => {
  const store = configureTestStore(analysesByProject);
  expect(store.getState()).toMatchSnapshot();

  store.dispatch(receiveProjectActivity(PROJECT_FOO, ANALYSES_FOO, PAGING));
  expect(store.getState()).toMatchSnapshot();

  store.dispatch(receiveProjectActivity(PROJECT_FOO, ANALYSES_FOO_2, { pageIndex: 2 }));
  expect(store.getState()).toMatchSnapshot();

  store.dispatch(receiveProjectActivity(PROJECT_BAR, ANALYSES_BAR, PAGING));
  expect(store.getState()).toMatchSnapshot();

  store.dispatch(deleteAnalysis(PROJECT_FOO, 'AVgFqeOSKpGuA48ADATE'));
  expect(store.getState()).toMatchSnapshot();

  store.dispatch(deleteAnalysis(PROJECT_BAR, 'AVgGkRvCrrTJiPpCD-rG'));
  expect(store.getState()).toMatchSnapshot();
});
