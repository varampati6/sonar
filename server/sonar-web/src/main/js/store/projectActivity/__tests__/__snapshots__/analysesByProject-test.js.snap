// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer 1`] = `Object {}`;

exports[`reducer 2`] = `
Object {
  "project-foo": Array [
    "AVgFqeOSKpGuA48ADATE",
    "AVgAgC1Vdo07z3PUnnkt",
  ],
}
`;

exports[`reducer 3`] = `
Object {
  "project-foo": Array [
    "AVgFqeOSKpGuA48ADATE",
    "AVgAgC1Vdo07z3PUnnkt",
    "AVgFqeOSKpGuA48ADATX",
  ],
}
`;

exports[`reducer 4`] = `
Object {
  "project-bar": Array [
    "AVgGkRvCrrTJiPpCD-rG",
  ],
  "project-foo": Array [
    "AVgFqeOSKpGuA48ADATE",
    "AVgAgC1Vdo07z3PUnnkt",
    "AVgFqeOSKpGuA48ADATX",
  ],
}
`;

exports[`reducer 5`] = `
Object {
  "project-bar": Array [
    "AVgGkRvCrrTJiPpCD-rG",
  ],
  "project-foo": Array [
    "AVgAgC1Vdo07z3PUnnkt",
    "AVgFqeOSKpGuA48ADATX",
  ],
}
`;

exports[`reducer 6`] = `
Object {
  "project-bar": Array [],
  "project-foo": Array [
    "AVgAgC1Vdo07z3PUnnkt",
    "AVgFqeOSKpGuA48ADATX",
  ],
}
`;
